### ISSUE TITLE
TLS support for BuildMaster <--> BuildSlave

### ISSUE THREAD

**Issue by tardyp** [buildbot/buildbot#2866]
This ticket is a migrated Trac ticket [2880](http://trac.buildbot.net/ticket/2880)

People contributed to the original ticket: @djmitche, @sa2ajj, @unknown_contributor
Ticket created on: `Aug 28 2014`
Ticket last modified on: `Aug 30 2016`

---

Currently BuildMaster and BuildSlave communicate over plain TCP/IP.

Some people find it rather risky to use Buildbot.

A few links to the relevant Twisted Python documentation:
* [[Getting Connected With Endpoints|http://twistedmatrix.com/documents/current/core/howto/endpoints.html]]
* [[server[[FromString]] function|https://twistedmatrix.com/documents/14.0.0/api/twisted.internet.endpoints.server[[FromString]].html]]
* [[client[[FromString]] function|https://twistedmatrix.com/documents/14.0.0/api/twisted.internet.endpoints.client[[FromString]].html]]

---

_Comment from_: @sa2ajj
_Date_: `Aug 28 2014`

Related ticket #2129

---
_Comment from_: @dustin
_Date_: `Aug 30 2014`

I expect this is possible with correct configuration, but I haven't tried that, nor do I know offhand what that congiguration is.

So this may be as simple as a documentation fix.

---
_Comment from_: @dustin
_Date_: `Sep 13 2014`

A brief bit of experimentation shows that on the master,

    c['slavePortnum'] = ''ssl:9999:privateKey=/tmp/key.pem:certKey=/tmp/crt.pem'

works.

However, the slave is not so lucky -- it instantiates an `internet.TCPClient` directly, using separate host and port parameters.

So this will need some code change on the slave in order to support TLS connections.

---
_Comment from_: @unknown_contributor
_Date_: `Aug 30 2016`

Working on it this week !

**Comment by seankelly** [comment]
Still open PR for this is at #2387.

**Comment by myfreeweb** [comment]
Temporary simple hack:

```diff
--- pb.py.orig	2018-11-11 13:33:02.490246000 +0000
+++ pb.py	2018-11-11 13:58:19.837791000 +0000
@@ -25,6 +25,7 @@
 from twisted.internet import error
 from twisted.internet import reactor
 from twisted.internet import task
+from twisted.internet import ssl
 from twisted.python import log
 from twisted.spread import pb
 
@@ -199,9 +200,11 @@
             maxRetries=self.maxRetries, maxRetriesCallback=self.gracefulShutdown)
         bf.startLogin(
             credentials.UsernamePassword(name, passwd), client=self.bot)
-        self.connection = c = internet.TCPClient(
+        self.connection = c = internet.SSLClient(
             buildmaster_host, port,
-            HangCheckFactory(bf, hung_callback=self._hung_connection))
+            HangCheckFactory(bf, hung_callback=self._hung_connection),
+            ssl.ClientContextFactory()
+            )
         c.setServiceParent(self)
 
     def _hung_connection(self):
```

Or as a monkey patch in `buildbot.tac`:

```python
from twisted.application import internet
from twisted.internet import ssl

def make_tls_client(h, p, f):
    tls_ctxf = ssl.ClientContextFactory()
    return internet.SSLClient(h, p, f, tls_ctxf)

internet.TCPClient = make_tls_client
```

**Comment by seankelly** [comment]
This is fixed from #4609. The PR added an integration test using TLS.

**Comment by rodrigc** [comment]
Documentation for this new feature is here:

http://docs.buildbot.net/latest/manual/configuration/global.html#setting-the-pb-port-for-workers

http://docs.buildbot.net/latest/manual/installation/worker.html#worker-tls-config
